src/FnTools/Types/Result.cs

```diff
@@ -0,0 +1,264 @@
+using System;
+using System.Collections.Generic;
+using FnTools.Exceptions;
+using FnTools.Types.Interfaces;
+using static FnTools.Prelude;
+
+namespace FnTools.Types
+{
+    public readonly struct Result<TOk, TError> : IGettable<TOk>, IEquatable<Result<TOk, TError>>, IToOption<TOk>, IToEither<TOk>
+    {
+        private readonly TError _error;
+
+        private readonly TOk _ok;
+
+        public bool IsOk { get; }
+
+        public bool IsError => !IsOk;
+
+        public Option<TError> Error => IsError ? Some(_error) : None;
+
+        public Result(TOk ok)
+        {
+            _ok = ok;
+            _error = default;
+            IsOk = true;
+        }
+
+        public Result(TError error)
+        {
+            _error = error;
+            _ok = default;
+            IsOk = false;
+        }
+
+        public bool Exists(Func<TOk, bool> condition)
+        {
+            _ = condition ?? throw new ArgumentNullException(nameof(condition));
+
+            return IsOk && condition(_ok);
+        }
+
+        public Result<T, TError> Map<T>(Func<TOk, T> map)
+        {
+            _ = map ?? throw new ArgumentNullException(nameof(map));
+
+            return IsOk ? new Result<T, TError>(map(_ok)) : new Result<T, TError>(_error);
+        }
+
+        public Result<TTOk, TTError> BiMap<TTOk, TTError>(Func<TOk, TTOk> map, Func<TError, TTError> errorMap)
+        {
+            _ = map ?? throw new ArgumentNullException(nameof(map));
+            _ = errorMap ?? throw new ArgumentNullException(nameof(errorMap));
+
+            return IsOk ? Ok<TTOk, TTError>(map(_ok)) : Error<TTOk, TTError>(errorMap(_error));
+        }
+
+        public Result<TOk, T> ErrorMap<T>(Func<TError, T> map)
+        {
+            _ = map ?? throw new ArgumentNullException(nameof(map));
+
+            return IsOk ? new Result<TOk, T>(_ok) : new Result<TOk, T>(map(_error));
+        }
+
+        public Result<T, TError> FlatMap<T>(Func<TOk, Result<T, TError>> flatMap)
+        {
+            _ = flatMap ?? throw new ArgumentNullException(nameof(flatMap));
+
+            return IsOk ? flatMap(_ok) : new Result<T, TError>(_error);
+        }
+
+        public Result<TOk, TError> FlatTap<T>(Func<TOk, Result<T, TError>> flatTap)
+        {
+            _ = flatTap ?? throw new ArgumentNullException(nameof(flatTap));
+
+            return FlatMap(x => flatTap(x).Map(_ => x));
+        }
+
+        public Result<TOk, TError> Recover(Func<TError, TOk> recover)
+        {
+            _ = recover ?? throw new ArgumentNullException(nameof(recover));
+
+            return IsOk ? this : new Result<TOk, TError>(recover(_error));
+        }
+
+        public Result<TOk, TError> RecoverWith(Func<TError, Result<TOk, TError>> recover)
```

Comment:
I think when you recover with a monadic function the error can change
```suggestion
        public Result<TOk, TErrorResult> RecoverWith(Func<TError, Result<TOk, TErrorResult>> recover)
```